examples/directchat.nim

```diff
@@ -0,0 +1,224 @@
+when not(compileOption("threads")):
+  {.fatal: "Please, compile this program with the --threads:on option!".}
+
+import tables, options, sequtils, algorithm, strformat, os, strutils
+import chronos
+import ../libp2p/switch,
+       ../libp2p/multistream,
+       ../libp2p/crypto/crypto,
+       ../libp2p/protocols/identify,
+       ../libp2p/connection,
+       ../libp2p/transports/[transport, tcptransport],
+       ../libp2p/multiaddress,
+       ../libp2p/peerinfo,
+       ../libp2p/peer,
+       ../libp2p/protocols/protocol,
+       ../libp2p/protocols/secure/secure,
+       ../libp2p/protocols/secure/secio,
+       ../libp2p/protocols/pubsub/pubsub,
+       ../libp2p/protocols/pubsub/floodsub,
+       ../libp2p/muxers/muxer,
+       ../libp2p/muxers/mplex/mplex,
+       ../libp2p/muxers/mplex/types
+
+const ChatCodec = "/nim-libp2p/chat/1.0.0"
+const DefaultAddr = "/ip4/127.0.0.1/tcp/55505"
+
+
+const Help = """
+  Commands: /[?|hep|connect|disconnect|exit]
+  help: Prints this help
+  connect: dials a remote peer
+  disconnect: ends current session
+  exit: closes the chat
+"""
+
+type
+  CustomData = ref object
+    consoleFd: AsyncFD
+    serveFut: Future[void]
+
+  ChatProto = ref object of LPProtocol
+    customData*: CustomData
+    switch: Switch
+    transp: StreamTransport
+    conn: Connection
+    client: bool
+    connected: bool
+    started: bool
+
+proc id (p: ChatProto): string =
+  result = "unknown"
+  if p.conn.peerInfo.peerId.isSome:
+    result = $p.conn.peerInfo.peerId.get()
+
+# forward declaration
+proc readWriteLoop(p: ChatProto) {.async, gcsafe.}
+proc readAndPrint(p: ChatProto) {.async, gcsafe.} =
+  while true:
+    while p.connected:
+      # echo &"{p.id} -> "
+      echo cast[string](await p.conn.readLp())
+    await sleepAsync(100.millis)
+
+proc dialPeer(p: ChatProto, address: string) {.async, gcsafe.} =
+  var parts = address.split("/")
+  if parts.len == 11 and parts[^2] != "ipfs" and parts[^2] != "p2p":
+    quit("invalid or incompelete peerId")
+
+  var remotePeer: PeerInfo
```

Comment:
prefer object init syntax + `let`:
```
let remotePeer = PeerInfo(
  peerId: some...
  adds: @[...]
)
```
this is less error-prone overall

Reply:
will be addressed in upcoming (more thorough) cleanup